packages/core/src/lib/implementation/compare-scorables.unit.test.ts

```diff
@@ -0,0 +1,320 @@
+            { slug: 'tests', title: 'Tests', score: 0.66 },
+            { slug: 'code', title: 'Code quality', score: 0.88 },
+          ],
+        },
+        after: {
+          categories: [
+            { slug: 'perf', title: 'Performance', score: 0.32 }, // regressed
+            { slug: 'tests', title: 'Tests passing', score: 0.69 }, // renamed + improved
+            { slug: 'code', title: 'Code quality', score: 0.88 },
+            { slug: 'security', title: 'Security', score: 1 }, // added
+          ],
+        },
+      } as ReportsToCompare),
+    ).toEqual<ReportsDiff['categories']>({
+      changed: [
+        {
+          slug: 'perf',
+          title: 'Performance',
+          scores: { before: 0.42, after: 0.32, diff: expect.closeTo(-0.1) },
+        },
+        {
+          slug: 'tests',
+          title: 'Tests passing',
+          scores: { before: 0.66, after: 0.69, diff: expect.closeTo(0.03) },
+        },
+      ],
+      unchanged: [{ slug: 'code', title: 'Code quality', score: 0.88 }],
+      added: [{ slug: 'security', title: 'Security', score: 1 }],
+      removed: [],
+    });
+  });
+});
+
+describe('compareGroups', () => {
+  it('should match groups by plugin-scoped slug and check for equal scores', () => {
+    expect(
+      compareGroups({
+        before: {
+          plugins: [
+            {
+              slug: 'eslint',
+              title: 'ESLint',
+              groups: [
+                { slug: 'problems', title: 'Problems', score: 0.4 },
+                { slug: 'suggestions', title: 'Suggestions', score: 0.5 },
+                { slug: 'formatting', title: 'Formatting', score: 0.9 },
+              ],
+            },
+            {
+              slug: 'stylelint',
+              title: 'Stylelint',
+              groups: [
+                { slug: 'errors', title: 'Avoid errors', score: 1 },
+                { slug: 'conventions', title: 'Conventions', score: 0.7 },
+                { slug: 'formatting', title: 'Whitespace etc.', score: 0.8 },
+              ],
+            },
+            {
+              slug: 'prettier',
+              title: 'Prettier',
+              groups: [
+                { slug: 'formatting', title: 'Code formatting', score: 0 },
+              ],
+            },
+          ],
+        },
+        after: {
+          plugins: [
+            {
+              slug: 'eslint',
+              title: 'ESLint',
+              groups: [
+                { slug: 'problems', title: 'Problems', score: 0.4 },
+                { slug: 'suggestions', title: 'Suggestions', score: 0.4 }, // regressed
+                { slug: 'formatting', title: 'Formatting', score: 0.95 }, // improved
+              ],
+            },
+            {
+              slug: 'stylelint',
+              title: 'Stylelint',
+              groups: [
+                { slug: 'errors', title: 'Avoid errors', score: 1 },
+                { slug: 'conventions', title: 'Conventions', score: 0.7 },
```

Comment:
As a personal preference, I would minimise the size of test data when the same result in logic is expected in multiple places (e.g. here for `errors` and `conventions`). It simplifies debugging.

I understand you wanted to provide realistic scenarios, just wondering if all cases couldn't have been tested using two plugins instead of three, so that the data fits on the screen without scrolling.

Reply:
OK, I omitted the Stylelint plugin :cry: